### DISCUSSION TITLE
Documentation update (Waveform/WaveformIn)

### DISCUSSION THREAD

**Discussion by aswyatt** [dls-controls/pythonSoftIOC#116] [General]
Could the documentation be updated to include WaveformIn - it took me a little while (evening through the code) to find out is an alias for Waveform. I presume it's to keep consistency with the other input types, and the Waveform type is for backwards compatibility. I personally will use WaveformIn since it makes the code more consistent and obvious. It was also a little for confusing as to why Waveform breaks the mould.

**Comment by Araneidae** [comment]
Yes, this is a good point, and it would make sense to update the documentation.

On the whole the `Waveform` name is indeed for backwards compatibility, but the reason it is that way is that there is a fundamental difference between the `waveform` record type and other record types (eg `ai` and `ao`).  Specifically, EPICS has dedicated and separate record types for "input" and "output" records (though the direction is terribly confusing, as `ai` records are for outputs from the IOC, and `ao` records are for control inputs to the IOC!)

However, there is no such distinction for the `waveform` record, instead it is up to the "device support layer" to decide in which direction the data flows.  In particular, when processing a `waveform` record, the record processing function can choose whether to read a freshly updated waveform from the record (treating as an "out" record) or to update the waveform in the record (making it an "in" record).  Indeed in principle a waveform record can dynamically switch between these two behaviours (though pythonSoftIoc does not support this)!

**Comment by Araneidae** [comment]
Fixed in 5f02a8452f85182a7ce9e003f32ef3b5afe4178f
